### ISSUE TITLE
Compiled java classes cannot be modifed before they are written to disk

### ISSUE THREAD

**Issue by steffschl** [eclipse-jdt/eclipse.jdt.core#983]
### Issue
I'm currently working on a project requiring to manipulate the byte code of compiled java classes. Currently this is achieved by hooking into 
```java
org.eclipse.jdt.core.compiler.CompilationParticipant.buildFinished(IJavaProject)
``` 
, reading the compiled classes from disk, applying the byte code modification and writing the resulting byte code back to disk.

This works, but introduces quite some extra runtime because of the additional read + write operations. This may be acceptable if just a few classes are to be modified, but for larger projects this quickly becomes a serious performance bottleneck, especially if the workspace is located on a network file system (NFS).

### Proposal

For the default java compiler there are compiler extensions available which allow to hook into the compile framework and manipulate the created java classes **before** they are written to disk, eliminating additional read+write operations. For the eclipse java compiler such possibilities are not available.

My proposal would be to introduce a new hook in
```java
org.eclipse.jdt.core.compiler.CompilationParticipant
```
which gives access to the compiled byte stream (read+write access) and some meta information (e.g. the BuildContext) right before it is written to disk, allowing to perform arbitrary byte code manipulations as required by the developer.

For example this new hook could look like this:
```java
public byte[] postProcess(BuildContext file, byte[] bytes);
```

This would enhance the possibilities to hook into the compile process and provide a general approach to modify the compiled byte code in performance-critical applications.

**Comment by szarnekow** [comment]
Is this the same as https://github.com/eclipse-jdt/eclipse.jdt.core/pull/694 ? Could you please review that part to check if it fits with your usecase?

**Comment by iloveeclipse** [comment]
> Is this the same as #694 ? Could you please review that part to check if it fits with your usecase?

If I read it right, #694 is about to save the bytes **generated by compiler** into the context, but not about modifying them.

**This** request is about possibility to **post process** class bytes generated by compiler **before** saving to disk, so at the end Eclipse writes the class file for each class **only once**.

**Comment by szarnekow** [comment]
> > Is this the same as #694 ? Could you please review that part to check if it fits with your usecase?
> 
> If I read it right, #694 is about to save the bytes **generated by compiler** into the context, but not about modifying them.

https://github.com/eclipse-jdt/eclipse.jdt.core/pull/694/files#diff-d80ea0327d9e01b92ff626d6dddb1df24781734eea3c9ceb0ddc14386d8c2947R42

I think this is covered by the setBytes(..) method on the ClassContent

**Comment by iloveeclipse** [comment]
> I think this is covered by the setBytes(..) method on the ClassContent

Not really, as the content is set from: `classContent.setBytes(classFile.getBytes());` - means, participants are receiving in the `ClassContent` the bytes that compiler generates, but they aren't asked to change the `classFile` itself.

This is what requested here - "class file weaving on the fly" **before** writing bytes to the disk, to avoid "generate class bytes -> write bytes to disk -> read bytes from disk -> modify class by participants > write class again" sequence.

The proposal in #694 would allow to skip `read bytes from disk` step and change 

`generate class bytes -> write bytes to disk -> read bytes from disk -> modify class by participants > write class again`

to 

`generate class bytes -> write bytes to disk -> read bytes from ClassContent -> modify class by participants > write class again`

Here the request is to change the order and reduce it to something like:

`generate class bytes -> read bytes from ClassContent -> modify class by participants -> write bytes to disk`

**Comment by steffschl** [comment]
I've just pull-requested the latest draft.
